app/assets/javascripts/Components/Result/file_viewer.jsx

```diff
@@ -112,57 +112,34 @@ export class FileViewer extends React.Component {
     });
 
     this.setState({loading: true, url: null}, () => {
-      if (!this.props.selectedFileURL) {
-        fetch(
-          Routes.get_file_course_submission_path(this.props.course_id, this.props.submission_id, {
-            submission_file_id: submission_file_id,
-            force_text: force_text,
-          }),
-          {credentials: "include"}
-        )
-          .then(res => res.json())
-          .then(body => {
-            if (body.type === "image" || body.type === "pdf" || this.isNotebook(body.type)) {
-              this.setState({type: body.type}, () => {
-                this.setFileUrl(submission_file_id);
-              });
-            } else {
-              const content = JSON.parse(body.content).replace(/\r?\n/gm, "\n");
-              this.setState({
-                content: content,
-                type: body.type,
-                loading: false,
-              });
-            }
-          });
+      this.setFileUrl(submission_file_id);
```

Comment:
Similar to my comment on your other PR, this function calls `setState`, which isn't synchronous. As you probably noticed, in the current implementation of `setFileUrl`, all of the URL-creation code is synchronous _except_ the `heic`/`heif` image formats, which currently require a separate request to be made.

You can try doing the following:

- Move the heic/heif handling into the `ImageViewer` component, where it more logically belongs
- Change `this.setFileUrl` to `this.getFileUrl` and just (synchronously) return a URL, and then use that to set the initial url in the `setState` call on line 114
- Ideally set the `type` state in that initial call (L114) as well, so that you don't need the if branch code on Lines 122-124
- See also my comment on the plaintext/binary branch a bit further below